Data Update Script to migrate logo_svg contents to png file (Will be merged and deployed on 10/01)

Ridhwana opened the pull request:
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description
As per https://forem.team/engineering/creator-signup-and-onboarding-16p6, we've updated the logo upload process for creators as follows:
- We only accept PNG/JPG/JPEG to be uploaded as a logo.
- We no longer accept SVG due to security vulnerabilities.

Based on the above we need to convert existing Forem’s logos from SVG (i.e, the `logo_svg` markup) to PNG so that there is no action for Forem Creators. 

Please note that this is a best effort script and we try to cater for as many conditions/edge cases as possible but we are unable to cater for everything. Hence we will be giving creators the opportunity to change/tweak their logos and re-upload if the resolution, visibility has been altered during the conversion. In addition, we will remind them that they should view their logo on both dark and light theme.

There are some Forems (very few it seems, like DEV, forem.dev and vscodetips) who have added custom scripts or are using dynamic CSS variables in their SVG logo to alter the logo’s styles for our dark theme and light theme (an example is the DEV logo). We are not able to convert these logos accurately with our tools.

Hence, we do our best effort to first provide static values for these CSS variable (that we know will not work in dark theme) and for others we want this script to log the errors for any of the logos that fail to convert.

## Related Tickets & Documents
Closes https://github.com/forem/forem/issues/15620

## QA Instructions, Screenshots, Recordings

_Test a successful update:_ 
1. Open a rails console and set the value for `logo_svg` 
```
Settings::General.logo_svg = '<svg width="300" height="100" xmlns="http://www.w3.org/2000/svg">
    <path fill="#dedede" stroke="#555" stroke-width="2" d="M2 2h296v96H2z"/>
    <text x="50%" y="50%" font-size="18" text-anchor="middle" alignment-baseline="middle" font-family="monospace, sans-serif" fill="#555">
        300×100
    </text>
</svg>'
```

2. Make sure that the logo's are nil. Open a rails console and run:
```
Settings::General.original_logo = nil 
Settings::General.resized_logo = nil
```
3. Run the Data Update Script in your terminal with `data_updates:run`
4. Open the rails console and check that the values for `Settings::General.original_logo` and `Settings::General.resized_logo` have updated.
5. Go to `/public/uploads/logos` and you should see an original and resized png versions of the original svg. 


You can also test a logo with a css variable like `currentColor` or `var(--base-inverted)`  and watch it convert if you'd like.
### UI accessibility concerns?

None.

## Added/updated tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help 

Some insights into fixing the Travis tests:
Initially the MiniMagick conversion process is failing on Travis.
```
CarrierWave::ProcessingError:
       Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /home/travis/build/forem/forem/tmp/1640166684-774651839101422-2549-8379/300x100.svg -auto-orient /tmp/image_processing20211222-11351-73ymxc.png` failed with error:
       convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
       convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
     # ./spec/uploaders/logo_svg_uploader_spec.rb:39:in `block (3 levels) in <main>'
     # ./spec/[secure]_helper.rb:145:in `block (3 levels) in <main>'
     # ./spec/[secure]_helper.rb:145:in `block (2 levels) in <main>'
     # ./spec/[secure]_helper.rb:100:in `block (2 levels) in <main>'
     # ------------------
     # --- Caused by: ---
     # MiniMagick::Error:
     #   `convert /home/travis/build/forem/forem/tmp/1640166684-774651839101422-2549-8379/300x100.svg -auto-orient /tmp/image_processing20211222-11351-73ymxc.png` failed with error:
     #   convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
     #   convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
     #   ./spec/uploaders/logo_svg_uploader_spec.rb:39:in `block (3 levels) in <main>'

```

 I tried a couple of things thus far like:
- Making sure that the fixture file is found as per [Mac's comment](https://github.com/forem/forem/pull/15710#discussion_r773195512)
- Testing the functionality with the fixture file locally (I moved it temporarily into the uploads folder): 
```
svg_file = File.open('./public/uploads/300x100.svg')
logo_svg_uploader = LogoSvgUploader.new.tap do |uploader|
  uploader.store!(svg_file)
end
```
 and it worked.
- https://travis-ci.community/t/build-times-out-with-no-apparent-reason/5083 as per this [commit](https://github.com/forem/forem/pull/15710/commits/53d96d006af2b7b6ab90344a336c294abe54639e).
- https://stackoverflow.com/questions/41138404/how-to-install-newer-imagemagick-with-webp-support-in-travis-ci-container as per this [commit](https://github.com/forem/forem/pull/15710/commits/9835d466e66e0eebfdfcdfc0b889b40acc3b3453)

I suspect it's a Travis configuration issue or maybe version difference in Travis, but haven't been able to figure it out as yet and would appreciate any insights on it.

See more of the conversation here: https://forem-team.slack.com/archives/C7GE84DEJ/p1641305439053700

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

I've chatted to Jennie about alerting Forems about the following: 
Forems will have a week to preview their converted uploads to check for:
1. resolution 
2. visibility 
3. and determine if they'd like to twek/update their logo to work on both dark adn light theme

Pawel will be tweaking our logos as well. 

We will send out a forem.dev post once we have run the script.

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](gif_link)

Ridhwana commented on app/uploaders/base_uploader.rb:
```diff
@@ -26,6 +26,8 @@ def size_range
 
   # strip EXIF (and GPS) data
   def strip_exif
+    return if file.content_type.include?("svg")
```
There will be no exif data for an SVG

jeremyf replied:
(nonblocker): This seems like a good inline comment for the code.

Ridhwana replied:
✅

Ridhwana commented on app/uploaders/logo_svg_uploader.rb:
```diff
@@ -0,0 +1,47 @@
+  IMAGE_TYPE_ALLOWLIST = %i[svg png].freeze
+  CONTENT_TYPE_ALLOWLIST = %w[image/svg+xml image/png].freeze
+
+  process convert: "png"
```
We set a process callback, which will call that method any time a file is uploaded. Thereafter the MiniMagick convert method which changes the image encoding format to the given format.

Ridhwana commented on app/uploaders/logo_uploader.rb:
```diff
@@ -29,13 +29,13 @@ def content_type_allowlist
 
   def filename
     # random_string in the filename to avoid caching issues
-    "original_logo_#{random_string}.#{file.extension}" if original_filename
```
This is a bit of an unrelated change but part of the same feature, we're not using the `original_filename` as part of the new file name, hence there is no need to check for it.

Ridhwana commented on lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:
```diff
@@ -0,0 +1,25 @@
+      return if ::Settings::General.try(:logo_svg).blank? ||
+        (::Settings::General.try(:original_logo).present? && ::Settings::General.try(:resized_logo).present?)
+
+      temp_svg = Tempfile.new(["logo", ".svg"])
+      temp_svg.write(::Settings::General.logo_svg)
+      temp_svg.close
```
We create a temporary file with the SVG string

citizen428 replied:
The ["Good practices" sectiob](https://ruby-doc.org/stdlib-3.0.2/libdoc/tempfile/rdoc/Tempfile.html#class-Tempfile-label-Good+practices) of the `Tempfile` documentation encourages the following form instead:

```ruby
Tempfile.create("logo.svg") do |file|
   # ...do something with file... e.g.
  file.write(::Settings::General.logo_svg))
end
```

To quote the [docs](https://ruby-doc.org/stdlib-3.0.2/libdoc/tempfile/rdoc/Tempfile.html#method-c-open)

> The File object will be automatically closed and the temporary file is removed after the block terminates, releasing all resources that the block created. The call returns the value of the block.

Ridhwana commented on lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:
```diff
@@ -0,0 +1,25 @@
+    rescue StandardError => e
+      # If we can't convert the logo, then alert ourselves so that we can track the Forems
+      # whose logos could not be converted
+      Rails.logger.error("Could not convert logo_svg for #{Settings::Community.community_name} Forem to a PNG")
```
This is to ensure that we are logging the error, so that we can alert the Forems whose SVG's were not able to convert successfully.

Ridhwana commented on lib/data_update_scripts/20211208074428_migrate_logo_svg_data.rb:
```diff
@@ -0,0 +1,25 @@
+      # whose logos could not be converted
+      Rails.logger.error("Could not convert logo_svg for #{Settings::Community.community_name} Forem to a PNG")
+      Honeybadger.notify(e)
+      ForemStatsClient.increment("logo_svg.conversion")
```
(blocking): I'm not sure how to use the ForemStatsClient properly, any idea's here?

maestromac replied:
That's our metric wrapper. This line of code would keep count on how often `logo_svg.conversion` action happens. It would result in a graph format that can be seen on Datadog.

Ridhwana commented on spec/uploaders/logo_svg_uploader_spec.rb:
```diff
@@ -0,0 +1,56 @@
+    end
+  end
+
+  describe "processed images" do
```
(blocking): These tests are failing on Travis with a MiniMagick error: 
```
Failed to manipulate with MiniMagick, maybe it is not an image? Original Error: `convert /home/travis/build/forem/forem/tmp/1639126885-827978499212382-0004-7279/300x100.svg -auto-orient /tmp/image_processing20211210-10742-7n7kz8.png` failed with error:
       convert-im6.q16: unable to read font `helvetica' @ error/annotate.c/RenderFreetype/1338.
       convert-im6.q16: non-conforming drawing primitive definition `text' @ error/draw.c/RenderMVGContent/4300.
```

but passing locally.

maestromac replied:
Hmm, try changing the following context on rails helper
```
-  config.fixture_path = "#{::Rails.root}/spec/fixtures"
+  config.fixture_path = "#{::Rails.root}/spec/fixtures/files"
```

Ridhwana commented on .travis.yml:
```diff
@@ -43,7 +43,38 @@ env:
     - DISABLE_SPRING=true
+    - IMAGEMAGICK_VERSION: '7.0.3-10'
 before_install:
+  # Update PATH so that travis can find newer imagemagick
+  - export PATH=$HOME/opt/bin:$PATH
+
+  # Checks if Imagemagick is already sufficient version
+  # If not installs it from the sources
+  - convert -version | grep $IMAGEMAGICK_VERSION || {
+    export CORES=$(nproc) &&
+    echo "Using $CORES cores for compiling..." &&
+    cd /tmp &&
+    curl -O https://storage.googleapis.com/downloads.webmproject.org/releases/webp/libwebp-$LIBWEBP_VERSION.tar.gz &&
+    tar xvzf libwebp-$LIBWEBP_VERSION.tar.gz &&
+    cd libwebp-* &&
+    ./configure --prefix=$HOME/opt &&
+    make -j$CORES &&
+    make install -j$CORES &&
+    cd /tmp &&
+    curl -O https://www.imagemagick.org/download/ImageMagick-$IMAGEMAGICK_VERSION.tar.gz &&
+    tar xvzf ImageMagick-$IMAGEMAGICK_VERSION.tar.gz &&
+    cd ImageMagick-* &&
+    ./configure --prefix=$HOME/opt &&
+    make -j$CORES &&
+    make install -j$CORES &&
+    $HOME/opt/bin/magick -version | grep $IMAGEMAGICK_VERSION &&
+    cd $TRAVIS_BUILD_DIR; }
+
+  # Update library paths for programs
+  - export LD_FLAGS=-L$HOME/opt/lib
+  - export LD_LIBRARY_PATH=/lib:/usr/lib:/usr/local/lib:$HOME/opt/lib
+  - export CPATH=$CPATH:$HOME/opt/include
+
```
I butchered the Travis configuration, but I would like to set a specific version of ImageMagick as per this [comment](https://stackoverflow.com/questions/41138404/how-to-install-newer-imagemagick-with-webp-support-in-travis-ci-container) and see if that helps.

djuber commented:
testing with a debug build suggests the errors in convert/minimagick/carrierwave are resolved by installing the `gsfonts` package in the image, before running rpsec.

https://forem-team.slack.com/archives/C7GE84DEJ/p1641317978064000?thread_ts=1641305439.053700&cid=C7GE84DEJ

Ridhwana commented on .travis.yml:
```diff
@@ -46,6 +46,8 @@ env:
     - DISABLE_SPRING=true
 before_install:
+  - sudo apt-get update
+  - sudo apt-get -y install gsfonts
```
We needed to add this so that the Travis build would pass. Thank you Dan!

Ridhwana commented on lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:
```diff
@@ -0,0 +1,30 @@
+
+      Rails.logger.error("Could not convert logo_svg for #{Settings::Community.community_name} Forem to a PNG.")
+
+      Honeybadger.notify(e, context: {
```
I would love a second pair of eyes on this from someone that is familiar with Honeybadger. 

We are expecting some logo conversions to fail and we want to be able to easily identify the failures in Honeybadger and notify the Forems that they need to do the conversion manually. Hence I've added the context and then a tag to easily filter these errors.

jeremyf replied:
@jgaskins might be a good reviewer of this

Ridhwana replied:
Thanks Jeremy, I think Jamie is on PTO but I've requested his review as well.

Ridhwana commented:
@citizen428 I made some changes and hence re-requesting your review.

jeremyf commented on app/uploaders/logo_svg_uploader.rb:
```diff
@@ -0,0 +1,47 @@
+  private
+
+  def random_string
+    SecureRandom.alphanumeric(20)
```
(non-blocking) Do we want to memoize this?  Wondering if we want `original_logo_asdfasdf.png` and `resized_logo_asdfasdf.png`?

Ridhwana replied:
I mean I don't think it would hurt, but maybe not a priority since this most likely a once-off script that we're running for existing Forems to migrate their existing logos. 

I could go in and optimise this on LogoUploader as that is used for our normal upload process, I'll go in and do this in a subsequent PR.

Ridhwana commented on lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:
```diff
@@ -0,0 +1,38 @@
+
+      # We do our best effort here: These are some (possibly not all) of the css variables that we may see in Forem
+      # logos. Hence, we try and replace them before converting.
+      logo_svg.gsub! "currentColor", "#090909"
+      logo_svg.gsub! "var(--base-inverted)", "#fff"
+      logo_svg.gsub! "var(--link-color)", "#3d3d3d"
```
Please note that this is a best effort script and we try to cater for as many cases where the user may have hacked their logo with CSS variable to work differently within the light and dark theme. These css variables will fall away and are converted to static values. 

However, we are unable to cater for all cases and we just replacing the most common css variables. 

We will be giving creators the opportunity to change/tweak their logos before launch.

msarit reviewed (commented):
@Ridhwana Hey there!

The happy path works as expected, but the DataUpdateScript for the failure path is succeeding (when it should fail) 🤔  I was sure to follow your excellent (🎉 ) QA instructions.

Ridhwana commented:
> @Ridhwana Hey there!
> 
> The happy path works as expected, but the DataUpdateScript for the failure path is succeeding (when it should fail) 🤔 I was sure to follow your excellent (🎉 ) QA instructions.

Oops.... I forgot to update the QA instructions when I added some [best effort updates to help convert logos that convert CSS variables](https://github.com/forem/forem/pull/15710#discussion_r779644145). Sorry Arit!
With the update referenced above (i.e. replacing the css variables with a static value)  we no longer need to test this as it will pass like you observed.

maestromac commented on app/uploaders/logo_svg_uploader.rb:
```diff
@@ -0,0 +1,65 @@
+        convert << temp_png_file.path
+      end
+
+      # https://github.com/minimagick/minimagick/issues/332
```
Did you mean to put this comment on line 39 instead?

maestromac reviewed (approved):
:clap: worked beautifully locally!

But to reiterate, once merged this job will definitely run on DEV and it will definitely fail?

jgaskins commented on lib/data_update_scripts/20220105112823_migrate_logo_svg_data.rb:
```diff
@@ -0,0 +1,30 @@
+
+      Rails.logger.error("Could not convert logo_svg for #{Settings::Community.community_name} Forem to a PNG.")
+
+      Honeybadger.notify(e, context: {
+                           community_name: Settings::Community.community_name,
+                           app_domain: Settings::General.app_domain,
+                           tags: "failed_svg_conversion"
+                         })
```
Only suggestion here is to move `tags` outside of `context` since `tags` is a [top-level property](https://www.rubydoc.info/gems/honeybadger/Honeybadger/Agent#notify-instance_method) of the Honeybadger notice. That should make it easier to search for.

```suggestion
      Honeybadger.notify(e, context: {
                           community_name: Settings::Community.community_name,
                           app_domain: Settings::General.app_domain,
                         },
                         tags: "failed_svg_conversion")
```

jgaskins reviewed (approved):
Just one suggestion. Beyond that, this looks great!

The search you'll want to use on Honeybadger is [this one](https://app.honeybadger.io/projects/72638/faults?sort=last_seen_desc&q=tag%3A%22failed_svg_conversion%22%E2%80%81-is%3Aresolved%E2%80%81-is%3Aignored%E2%80%81). Also, thank you so much for your attention there! :100:

juliannatetreault reviewed (approved):
Let's do this! 🚀

Ridhwana commented:
> 👏 worked beautifully locally!
> 
> But to reiterate, once merged this job will definitely run on DEV and it will definitely fail?

@maestromac this should not fail on DEV locally anymore, and I've tested this. We changed track based on this [curveball](https://forem-team.slack.com/archives/C01S4QZUDNC/p1641461142000400?thread_ts=1641403412.444700&cid=C01S4QZUDNC) and we added [some functionality to replace the common CSS variables with static values](https://github.com/forem/forem/pull/15710/files#r779644145). This will help Forems like DEV.to and Forem.dev to pass. If all goes well with this change we expect most Forems to pass because after looking at the data it seems to be mostly our Forems that use CSS Variables (which is now catered for in the script).

nickytonline reviewed (approved):
🚀
